### PR TITLE
Gateways refactor

### PR DESCRIPTION
**Pull request by cws1121**
##### SUMMARY
Includes:
Gateway fees: refactoring the special-case code into methods on the model,
Gateway fees: got rid of _get_multipart_count method (redundant),
Gateway fees: add_xxxx_gateway_fee methods are converted to commands and as such are included in migrations
Gateway fees: got rid of _GatewayChargeInfo named tuple (redundant)
Gateway fees Tests: refactoring the special-case code - creating new group of backends (backends with direct fees)
Gateway fees Tests: tests for infobip gareway added
Infobip and Turn models: all shared code (related to whatsapp) is extracted into  whatsapputils module
Infobip model: fixed bug with template messages labeled as "invalid user response"
SQLSMSBackend model: new class property - using_api_to_get_fees
"Manage whatsapp templates" page - handling exceptions in the "page context" section

##### FEATURE FLAG
/

##### RISK ASSESSMENT / QA PLAN
Everything tested on local, should be redeployed and tested on staging as well.

##### PRODUCT DESCRIPTION
According to suggestions that I got during the code review of my previous PR:
https://github.com/dimagi/commcare-hq/pull/27845

It would be misleading to push this refactoring work to branch called aws_pinpoint_fees, that's why I've create a new PR.
(In the meantime I got instructions to revert all code related to Pinpoint fees. so that PR should be closed and branch should be deleted)

### CODE REVIEW

File: corehq/apps/smsbillables/management/commands/bootstrap_infobip_gateway.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
+import settings
```
Comment (millerdev) [inline corehq/apps/smsbillables/management/commands/bootstrap_infobip_gateway.py:1]: There are three new management commands in this PR with a fair amount of duplication. What do you think of combining them into a single management command with a parameter that specifies the gateway to bootstrap?

File: corehq/apps/sms/models.py [hunk 2]
```diff
@@ -2396,6 +2397,13 @@ def get_sms_rate_limit(self):
     def send(self, msg, *args, **kwargs):
         raise NotImplementedError("Please implement this method.")
 
+    @classproperty
+    def using_api_to_get_fees(cls):
+        """
+        Override in case backend is fetching gateway fees through provider API
+        """
+        return False
```
Comment (millerdev) [inline corehq/apps/sms/models.py:2405]: Why `classproperty` instead of a class attribute?

```suggestion
    # Override in case backend is fetching gateway fees through provider API
    using_api_to_get_fees = False
```

File: corehq/apps/smsbillables/models.py [hunk 3]
```diff
@@ -358,62 +338,78 @@ def create(cls, message_log, multipart_count=1):
...
+                        backend_api_id,
+                        direction,
+                    )
+                else:
+                    log_smsbillables_error(
+                        "Could not create gateway fee for message %s: no backend_message_id" % couch_id
+                    )
             else:
-                gateway_fee = SmsGatewayFee.get_by_criteria(
+                fixed_gateway_fee = SmsGatewayFee.get_by_criteria(
```
Comment (millerdev) [inline corehq/apps/smsbillables/models.py:371]: Why was this renamed? The new name does not seem more informative to me.

File: corehq/messaging/smsbackends/infobip/models.py [hunk 4]
```diff
@@ -20,6 +25,10 @@ class Meta(object):
         app_label = 'sms'
         proxy = True
 
+    @classproperty
+    def using_api_to_get_fees(cls):
+        return True
```
Comment (millerdev) [inline corehq/messaging/smsbackends/infobip/models.py:30]:
```suggestion
    using_api_to_get_fees = True
```

File: corehq/messaging/smsbackends/infobip/models.py [hunk 5]
```diff
@@ -145,5 +157,41 @@ def generate_template_string(cls, template):
...
+            return messages[0]
+        except Exception as e:
+            raise RetryBillableTaskException(str(e))
+
+    def _get_message_details(self, api_channel, api_suffix, config, headers, parameters):
+        url = f'https://{config.personalized_subdomain}.{INFOBIP_DOMAIN}{api_channel}{api_suffix}'
+        response = requests.get(url, params=parameters, headers=headers)
+        response_content = json.loads(response.content)
+        messages = response_content['results']
+        return messages
```
Comment (millerdev) [inline corehq/messaging/smsbackends/infobip/models.py:192]: requests has built-in support for JSON data
```suggestion
        return response.json()['results']
```

File: corehq/messaging/smsbackends/twilio/models.py [hunk 6]
```diff
@@ -26,6 +28,10 @@ class Meta(object):
         app_label = 'sms'
         proxy = True
 
+    @classproperty
+    def using_api_to_get_fees(cls):
+        return True
```
Comment (millerdev) [inline corehq/messaging/smsbackends/twilio/models.py:33]:
```suggestion
    using_api_to_get_fees = True
```

**Review by millerdev — COMMENTED** [review]
Great! Now we have abstraction that any gateway can implement to support pass-through fees. Thank you!

File: corehq/apps/smsbillables/management/commands/bootstrap_infobip_gateway.py (hunk 1 above)
Comment (nemanja1040) [reply corehq/apps/smsbillables/management/commands/bootstrap_infobip_gateway.py]: alright, I've created a new command called "bootstrap gateway fees". Similarly, other bootstrap gateway commands could be included if needed.

File: corehq/apps/smsbillables/models.py (hunk 3 above)
Comment (nemanja1040) [reply corehq/apps/smsbillables/models.py:372]: Well, judging by the names I assumed direct_gateway_fee is a component of gateway_fee but that is not the case. There is no correlation between these two. Also term "Fixed" fee might not be the best choice here. gateway_fee indicates fixed (or flat) fee that is being calculated based on other tables and documents that can be found in cchq and direct fee is fetched from API endpoints. On second thought, since SmsBillable has two fields called gateway_fee and direct_gateway_fee maybe it would be better call to rename fixed_gateway_fee back to gateway_fee. Should I do that?

Comment (millerdev) [reply corehq/apps/smsbillables/models.py:372]: I think that makes sense, yes.

**Review by millerdev — APPROVED** [review]
Thanks!
